[flang][openmp]Add UserReductionDetails and use in DECLARE REDUCTION

Leporacanthicus opened the pull request:
This adds another puzzle piece for the support of OpenMP DECLARE REDUCTION functionality.

This adds support for operators with derived types, as well as declaring multiple different types with the same name or operator.

A new detail class for UserReductionDetials is introduced to hold the list of types supported for a given reduction declaration.

Tests for parsing and symbol generation added.

Declare reduction is still not supported to lowering, it will generate a "Not yet implemented" fatal error.

Leporacanthicus committed:
[flang][openmp]Add UserReductionDetails and use in DECLARE REDUCTION

This adds another puzzle piece for the support of OpenMP DECLARE
REDUCTION functionality.

This adds support for operators with derived types, as well as declaring
multiple different types with the same name or operator.

A new detail class for UserReductionDetials is introduced to hold
the list of types supported for a given reduction declaration.

Tests for parsing and symbol generation added.

Declare reduction is still not supported to lowering, it
will generate a "Not yet implemented" fatal error.

tblah commented on flang/include/flang/Semantics/symbol.h:
```diff
@@ -701,14 +701,33 @@ class GenericDetails {
+  const TypeVector &GetTypeList() const { return typeList_; }
+
+  bool SupportsType(const DeclTypeSpec *type) const {
+    for (auto t : typeList_)
+      if (t == type)
+        return true;
+    return false;
```
nit: I think semantics style requires braces here

Leporacanthicus replied:
Done

tblah commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,15 +1748,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
   PopScope();
 }
 
+parser::CharBlock MakeNameFromOperator(
```
nit: could this be static? (and below)

Leporacanthicus replied:
No, both these functions are used in check-omp-structure.cpp too. That's why they are declared in resole-names-tools.h.

tblah commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,15 +1748,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+}
+
+parser::CharBlock MangleSpecialFunctions(const parser::CharBlock name) {
+  if (name == "max") {
+    return parser::CharBlock{"op.max", 6};
+  }
+  if (name == "min") {
+    return parser::CharBlock{"op.min", 6};
+  }
+  if (name == "iand") {
+    return parser::CharBlock{"op.iand", 7};
+  }
+  if (name == "ior") {
+    return parser::CharBlock{"op.ior", 6};
+  }
+  if (name == "ieor") {
+    return parser::CharBlock{"op.ieor", 7};
+  }
+  // All other names: return as is.
+  return name;
```
nit: llvm::StringSwitch might be more efficient here

Leporacanthicus replied:
Done.

tblah commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,15 +1748,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
     const std::optional<parser::OmpClauseList> &clauses) {
+  const parser::Name *name{nullptr};
+  parser::Name mangledName{};
+  UserReductionDetails reductionDetailsTemp{};
```
```suggestion
  UserReductionDetails reductionDetailsTemp;
```

Leporacanthicus replied:
Done. [and the line above too].

tblah commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,15 +1748,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+    name = &mangledName;
+  }
+
+  UserReductionDetails *reductionDetails{&reductionDetailsTemp};
```
what do we need the stack temporary for? You already check if it is null originally and I don't think it can be null in the second  case because you already checked that the symbol exists.

Leporacanthicus replied:
Adding a comment to explain the reason we do this.

tblah reviewed (commented):
Overall looks good, just a few minor comments. Good job thinking of all the edge cases to test.

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3464,9 +3466,11 @@ static bool IsReductionAllowedForType(
       case parser::DefinedOperator::IntrinsicOperator::Subtract:
-        return type.IsNumeric(TypeCategory::Integer) ||
+        if (type.IsNumeric(TypeCategory::Integer) ||
             type.IsNumeric(TypeCategory::Real) ||
-            type.IsNumeric(TypeCategory::Complex);
+            type.IsNumeric(TypeCategory::Complex))
+          return true;
+        break;
```
Is the same change required for the isLogical check below?

kiranchandramohan commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1798,6 +1862,13 @@ void OmpVisitor::ProcessReductionSpecifier(
+    if (!symbol) {
+      symbol = &MakeSymbol(mangledName, Attrs{}, std::move(*reductionDetails));
+    }
+    name->symbol = symbol;
```
It will be good to see a test for the multiple case as well. Does it override the symbol?

Leporacanthicus replied:
I'm adding a new test for dupliocate symbols. There is one, but I think we need another different one.

And yes, we need to set the symbol on the original name, even if the actual name is not matching, so updating the symbol is required. If not, another portion finds the name and says it's not been resolved.

kiranchandramohan reviewed (commented):
A few initial comments.

Leporacanthicus commented on flang/include/flang/Semantics/symbol.h:
```diff
@@ -701,14 +701,33 @@ class GenericDetails {
 };
 llvm::raw_ostream &operator<<(llvm::raw_ostream &, const GenericDetails &);
 
+class UserReductionDetails : public WithBindName {
```
Probably shouldn't inherit anything... :)

Leporacanthicus committed:
Fix review comments

* Add two more tests (multiple operator-based declarations and re-using
  symbol already declared.
* Add a few comments.
* Fix up logical results.

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,15 +1748,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+    return parser::CharBlock{"op.NEQV", 8};
+
+  default:
+    assert(0 && "Unsupported operator...");
```
Done.

Leporacanthicus commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3479,8 +3483,18 @@ static bool IsReductionAllowedForType(
+      return false;
     }
-    return true;
+    assert(0 && "Intrinsic Operator not found - parsing gone wrong?");
```
Done.

Leporacanthicus committed:
Use stringswitch and spell details correctly

Leporacanthicus commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3493,18 +3507,42 @@ static bool IsReductionAllowedForType(
+
+      // We also need to check for mangled names (max, min, iand, ieor and ior)
+      // and then check if the type is there.
+      parser::CharBlock mangledName = MangleSpecialFunctions(name->source);
```
Done

kiranchandramohan commented:
There is a case for UserReductionDetails that has to be added while writing a module file. The following test crashes in that code.
```
module decl_red1
  type t1
    integer :: val
  end type t1
  interface operator(*)
    module procedure mul
  end interface
!$omp declare reduction(*:t1:omp_out=omp_out*omp_in) initializer(omp_priv=t1(1))
contains
  type(t1) function mul(v1, v2)
    type(t1), intent (in):: v1, v2
    mul%val = v1%val * v2%val
  end function
end module
```

NimishMishra commented on flang/lib/Semantics/symbol.cpp:
```diff
@@ -246,7 +246,7 @@ void GenericDetails::CopyFrom(const GenericDetails &from) {
 std::string DetailsToString(const Details &details) {
   return common::visit(
-      common::visitors{
+      common::visitors{//
```
Stray "//" ?

Leporacanthicus replied:
No idea how that happened. I don't use VI. :)

NimishMishra commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,14 +1749,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+
+  default:
+    DIE("Unsupported operator...");
+    return parser::CharBlock{"op.?", 4};
```
Will `return parser::CharBlock{"op.?", 4};` be executed after the `DIE`?

Leporacanthicus replied:
As below (I'm working "up" the list), this was an assert, which of course gets removed in some builds. DIE, as I understand it, always kills the compiler process, regardless of build.

NimishMishra commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3493,18 +3510,42 @@ static bool IsReductionAllowedForType(
-    // TODO: user defined reduction operators. Just allow everything for now.
-    return true;
+    DIE("name and name->symbol should be set here...");
+    return false;
```
Will `return false;` be executed after the `DIE`?

Leporacanthicus replied:
No, this used to be an assert, and when changing to die, I didn't remove the line below.

NimishMishra commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3493,18 +3510,42 @@ static bool IsReductionAllowedForType(
     }
-    // TODO: user defined reduction operators. Just allow everything for now.
-    return true;
+    DIE("name and name->symbol should be set here...");
```
When we say `should be set here`, do we want to give a location for the error? Can `name->symbol` be rather wrapped as an assertion in this function rather than a `DIE`?

NimishMishra commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,14 +1749,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+    return parser::CharBlock{"op.NEQV", 8};
+
+  default:
+    DIE("Unsupported operator...");
```
Should we emit a semantic error for unsupported operator here? Is there a specific reason to prefer `DIE` over an error?

Leporacanthicus replied:
Good spot, I thought there was some other code that stops us from using "bad" combitnation here, but it turns out that you can do `!$omp declare reduction(/:integer:...)` and make it die. Error message printing implemented, and another test added.

NimishMishra reviewed (commented):
Thanks for the patch. A few nits.

Leporacanthicus committed:
Add support for user defined operators in declare reduction

Also print the reduction declaration in the module file.

Fix trivial typo.

Add/modify tests to cover all the new things, including fixing
the duplicated typo in the test...

Leporacanthicus commented:
> There is a case for UserReductionDetails that has to be added while writing a module file. The following test crashes in that code.
> 
> ```
> module decl_red1
>   type t1
>     integer :: val
>   end type t1
>   interface operator(*)
>     module procedure mul
>   end interface
> !$omp declare reduction(*:t1:omp_out=omp_out*omp_in) initializer(omp_priv=t1(1))
> contains
>   type(t1) function mul(v1, v2)
>     type(t1), intent (in):: v1, v2
>     mul%val = v1%val * v2%val
>   end function
> end module
> ```

Now works, and additional test added.

Leporacanthicus committed:
Fix nit comments and add simple bad operator test

Leporacanthicus committed:
Fix error messages to be more consistent

Leporacanthicus commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1748,14 +1749,75 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+  } else {
+    const auto &defOp{std::get<parser::DefinedOperator>(id.u)};
+    mangledName.source = MakeNameFromOperator(
+        std::get<parser::DefinedOperator::IntrinsicOperator>(defOp.u));
```
Now fixed, and additional tests added.

Leporacanthicus committed:
add missed test change

Leporacanthicus committed:
Improve support for metadirective + declare reduction

kiranchandramohan commented:
Adding @klausler for the changes in writing mod-file, and other areas like symbol.{cpp,h}.

klausler commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -1035,6 +1037,25 @@ void ModFileWriter::PutTypeParam(llvm::raw_ostream &os, const Symbol &symbol) {
+  // declaration. There may be multiple declarations.
+  // Decls are pointers, so do not use a referene.
+  for (const auto decl : details.GetDeclList()) {
+    if (auto d = std::get_if<const parser::OpenMPDeclareReductionConstruct *>(
```
Use `common::visit` on a lambda with a `const auto &` argument instead!

Leporacanthicus replied:
:+1:

Leporacanthicus committed:
Fix Klausler reported review comments

Also rebase, as the branch was quite a way behind. Small conflict
was resolved.

Leporacanthicus commented on flang/include/flang/Semantics/symbol.h:
```diff
@@ -701,14 +703,41 @@ class GenericDetails {
+  void AddType(const DeclTypeSpec *type) { typeList_.push_back(type); }
+  const TypeVector &GetTypeList() const { return typeList_; }
+
+  bool SupportsType(const DeclTypeSpec *type) const {
```
:+1:

Leporacanthicus commented on flang/include/flang/Semantics/symbol.h:
```diff
@@ -701,14 +703,41 @@ class GenericDetails {
+
+  UserReductionDetails() = default;
+
+  void AddType(const DeclTypeSpec *type) { typeList_.push_back(type); }
```
:+1:

Leporacanthicus commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3361,8 +3362,8 @@ bool OmpStructureChecker::CheckReductionOperator(
-        auto *misc{name->symbol->detailsIf<MiscDetails>()};
-        valid = misc && misc->kind() == MiscDetails::Kind::ConstructName;
+        auto *reductionDetails{name->symbol->detailsIf<UserReductionDetails>()};
+        valid = reductionDetails != nullptr;
```
:+1:

Leporacanthicus commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -1035,6 +1037,25 @@ void ModFileWriter::PutTypeParam(llvm::raw_ostream &os, const Symbol &symbol) {
 
+void ModFileWriter::PutUserReduction(
+    llvm::raw_ostream &os, const Symbol &symbol) {
+  auto &details{symbol.get<UserReductionDetails>()};
```
:+1:

kiranchandramohan commented:
Can you check whether the declare reduction works with renamed operators. I am giving an example below for reference, but please think through in general about the possibility of renamed operators.

```
module my_module
    implicit none

    type :: my_type
        real :: value
    end type my_type

    interface operator(.mul.)
        module procedure my_mul
    end interface

contains

    function my_mul(a, b) result(c)
        type(my_type), intent(in) :: a, b
        type(my_type) :: c
        c%value = a%value * b%value
    end function my_mul

end module my_module

!------------------------------

program test_omp_reduction
  use omp_lib
  use my_module, only: my_type, operator(.submul.) => operator(.mul.)
  implicit none

  type(my_type) :: result
  integer :: i
  !$omp declare reduction (.submul. : my_type : omp_out = omp_out .submul. omp_in) initializer(omp_priv = my_type(1.0))

  result = my_type(1.0)

  
  !$omp parallel do reduction(.submul.:result)
  do i = 1, 10
      result = result .submul. my_type(real(i))
  end do
  !$omp end parallel do

  print *, 'Final result: ', result%value
end program test_omp_reduction
```

Leporacanthicus committed:
Fix some semantics issues

Leporacanthicus committed:
[Flang][OpenMP] Fix review comment failed examples

Add code to better handle operators in parsing and semantics.
Add a function to set the the scope when processign assignments,
which caused a crash in "check for pure functions".

Add three new tests and amend existing tests to cover a pure function.

Leporacanthicus commented:
> There is a crash in `Fortran::semantics::IsReductionAllowedForType` for the following test.
> 
[snip big chunk of code]

> pure logical function is_reduction_valid()
>   !$omp declare reduction (bar : integer : omp_out = omp_out + omp_in) initializer (omp_priv = 0)
>   is_reduction_valid = .false.
> end function is_reduction_valid
> ```

Fixed all three of the code snippets that fail, and added tests for checking that it works.

kparzysz commented on flang/lib/Semantics/mod-file.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "mod-file.h"
 #include "resolve-names.h"
+#include "flang/Common/indirection.h"
```
Seems unnecessary...

Leporacanthicus replied:
Done!

kparzysz commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -1752,14 +1761,91 @@ void OmpVisitor::ProcessMapperSpecifier(const parser::OmpMapperSpecifier &spec,
+    reductionDetails = symbol->detailsIf<UserReductionDetails>();
+
+    if (!reductionDetails) {
+      context().Say(name->source,
```
The "source" argument to Say must point to somewhere within the cooked sources.  It's used to locate the line number and for underlining of the problematic fragment.  I'm not sure if that's guaranteed here, since some names are created within this function.

Leporacanthicus replied:
Done!

kparzysz commented on flang/include/flang/Semantics/semantics.h:
```diff
@@ -343,6 +347,11 @@ class SemanticsContext {
   UnorderedSymbolSet isDefined_;
   std::list<ProgramTree> programTrees_;
+
+  // storage for mangled names used in OMP DECLARE REDUCTION.
```
Typo: "storage" -> "Storage".

Leporacanthicus committed:
Fix review comments

Leporacanthicus commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3485,8 +3496,20 @@ void OmpStructureChecker::CheckReductionObjects(
 
+static bool CheckSymbolSupportsType(const Scope &scope,
+    const parser::CharBlock &name, const DeclTypeSpec &type) {
+  if (const auto &symbol{scope.FindSymbol(name)}) {
```
Done!

tblah commented:
Mats will no longer be working on this. Please comment on the new pull request at https://github.com/llvm/llvm-project/pull/140066

Apologies for the noise and loss of comments.
